BLSQ/iaso
# IA-2642: Payments part 1

Api to get potential payments and web interface (filters, order)

# `Potential payment` API

[Mock-ups](https://whimsical.com/supervision-of-the-users-before-payment-9rNU57j31PnFykziJLocbZ)

This API allows to list potential payments linked to multiple `OrgUnitChangeRequest` by the same user to be updated and queried.

The Django model that stores "Potential payment" is `PotentialPayment`.

Whenever the list endpoint is invoked, it evaluates whether a new change request can be incorporated into the potential payment, or if there's a need to generate a new potential payment.

# Get a `PotentialPayment` list

## Permissions

- User must be authenticated
- User needs `iaso_payments` permission


## Query Parameters - Filters

- `page`: Int (optional) - Specifies the current page number. If not provided, the default value is 1.
- `order`: String (optional) - Specifies the order in which the results should be returned. If not provided, the default ordering value is `id`.
- `limit`: Int (optional) - Defines the number of entities to be returned per page. The default value is 20 if not specified.
- `users`: String (optional) - A comma-separated list of User IDs associated with the payments
  - Example: `&user_ids=10,9`
- `user_roles`: String (optional) - A comma-separated list of User Role IDs associated with the payments
  - Example: `&user_role_ids=10,9`
- `parent_id`: Int (optional) - The ID of the parent organization unit linked to the change requests. This should also include child units.
- `change_requests__created_at_after`: Date - 'YYYY-MM-DD' (optional) - The start date for when the change request has been validated. 
- `change_requests__created_at_before`: Date - 'YYYY-MM-DD' (optional) - The end date for when the change request has been validated.


## Possible responses

### 200 - OK

```json
{
  "count": "Long",
  "has_next": "Boolean",
  "has_previous": "Boolean",
  "page": "Long",
  "pages": "Long",
  "limit": "Long",
  "results": [
    "id": "Int - unique id",
    "change_requests": [
      {
        "id": "Int - change request unique id",
        "org_unit_id": "String - id or UUID of the OrgUnit to change",
        "uuid": "String - change request unique id",
      }
    ],
    "user": {
        "id": "Int - user unique id",
        "username": "String - username of user",
        "first_name": "String - first name of user",
        "last_name": "String - last name of user",
        "telephone": "String? - This type of field should be specified - not implemented yes",
    },
  ]
}
```

### 400 - Bad request

- `page` or `limit` cannot be parsed to a correct integer value

### 401 - Unauthorized

- No authentication token or an invalid one was provided

### 403 - Forbidden

- User doesn't have the proper permission to access this resource.


### 404 - Not found

- `users`, `user_roles`, parent_id` not found

Related JIRA tickets : IA-2462

## Self proofreading checklist


## Doc

Api is documented on swagger [endpoint](http://localhost:8081/swagger/)

## Changes

- new model, api, serializer, filters to get potential payments
- front-end interface to display, filter and ordering potential payments
- back-end test

## How to test

Login with a superuser account or add `iaso_payments` permission to your user.

You need to have a few `VALIDATED` change request on your account (created_by field should be set)

Open the menu and navigate to Payments/To Pay 

Play with the results.


## Print screen / video


https://github.com/BLSQ/iaso/assets/12494624/a3b7d0d6-cf13-4ea1-810a-7dbd22b2710f



## Notes

This is the first part, `Lots` are not working for new
